packages/package-json-lint/src/rules/require-no-hard-coded-dependency-versions.js

```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  create: ({ ruleConfig, projectType, report }) => ({
+    dependencies: (dependencies) => {
+      const compatibleVersionRegexList = ['dev', 'latest', 'main', 'master', /(\^|<|<=|>|>=|~)\s{0,1}\d/, /^x\.|\.x\.|\.x$/, /[\d.]{0,2}\d\s{0,1}-\s{0,1}[\d.]{0,2}\d/];
+      if (projectType === 'module' || projectType === 'devModule') {
+        const messageString = 'require-no-hard-coded-dependency-versions';
+        const currentProblems = Object.keys(dependencies).map(dependencyName => {
+          const dependencyVersion = dependencies[dependencyName];
+          const isCompatibleVersion = compatibleVersionRegexList.map(versionRegex => (!!dependencyVersion.match(versionRegex))).includes(true);
```

Comment:
I did a little digging with semver. I wonder if we can say:

const isCompatibleVersion = semver.clean(version) === null;

Reply:
It doesn't look like [semver.clean](https://github.com/npm/node-semver#clean) can handle ranges so syntax list `">= 2.3.0-dev.0 <2.3.0-dev.x"` or even just `2.3.0-dev.x` would not work so we may still need to handle those separately.

Reply:
I think it’ll handle them fine. We want them to return undefined or null. Basically anything but a solitary version should be fine

Reply:
@benbcai,
`semver.clean` works for `>= 2.3.0-dev.0 <2.3.0-dev.x`. And it will work if it has `< >` symbols. 
It only doesn't work for `2.3.0-dev.x` as this seems to be a fixed version.

If we are sure about it's going to be `dev.x or branch.x`, we can add it as an exception and have a check for it using regex.
What do you think? @ryanthemanuel @benbcai

Reply:
2.3.0-dev.x will *only* resolve to a version that is released as 2.3.0-dev.x so this is effectively hard coded. I think it's fine to not allow this.

Reply:
e.g. 1.0.0-rc.x does not actually resolve to 1.0.0-rc.1 

From https://semver.npmjs.com/

![image](https://user-images.githubusercontent.com/4873279/134350689-3d867470-c8cc-4286-9bc6-34b6dbcf60a9.png)

Reply:
I confused myself at first but after thinking thru this again, it all makes sense now.